### PR TITLE
perf: add noexcept and const to proper functions

### PR DESCRIPTION
**Pull request by anonrig**

### CODE REVIEW

File: src/parser.cpp [hunk 1]
```diff
@@ -1081,7 +1077,7 @@ namespace ada::parser {
           // Otherwise, run these steps:
           else {
             // UTF-8 percent-encode c using the path percent-encode set and append the result to buffer.
-            buffer += unicode::utf8_percent_encode(std::string{*pointer}, character_sets::PATH_PERCENT_ENCODE);
+            buffer.append(unicode::utf8_percent_encode(std::string{*pointer}, character_sets::PATH_PERCENT_ENCODE));
```
Comment (lemire) [inline src/parser.cpp:1080]: So here, buffer.append might throw (in theory), if it does and you have declare the function noexcept, you may get a hard crash.

File: src/unicode.cpp [hunk 2]
```diff
@@ -108,9 +99,9 @@ namespace ada::unicode {
 
     for (uint8_t iterator: input) {
       if (character_sets::bit_at(character_set, iterator)) {
-        result += character_sets::hex + iterator * 4;
+        result.append(character_sets::hex + iterator * 4);
```
Comment (lemire) [inline src/unicode.cpp:102]: The same is true here, you are using standard C++ call that *may* throw, inside a noexcept function. That is a bit controversial.

**Comment by lemire** [comment]
Qualifiers such as constexpr, const, noexcept... are unlikely to affect the performance in my view (one can benchmark). 

Being able to declare a user-facing function constexpr is great because it means that the function can be computed at compile time...

File: src/parser.cpp (hunk 1 above)
Comment (anonrig) [reply src/parser.cpp:1080]: If I'm not mistaken, it only throws in the following case, [referencing docs](https://cplusplus.com/reference/string/string/append/), which is not the case for us.

```
If this is greater than str's [length](https://cplusplus.com/string::length), it throws [out_of_range](https://cplusplus.com/out_of_range).
```

File: src/unicode.cpp (hunk 2 above)
Comment (anonrig) [reply src/unicode.cpp:102]: Is there a performance benefit for using `append` instead of `+=`?

Comment (lemire) [reply src/unicode.cpp:102]: You'd want to use neither, honestly.

**Pull request closed without merging**